File: golang/78/78.go
package main

var res [][]int
var cur []int

func subsets(nums []int) [][]int {
	res = make([][]int, 0)
	cur = make([]int, 0)

	backtrack(nums, 0)

	return res
}

func backtrack(nums []int, start int) {
	t := make([]int, len(cur))
	copy(t, cur)
	res = append(res, t)

	for i := start; i < len(nums); i++ {
		cur = append(cur, nums[i])
		backtrack(nums, i+1)
		cur = cur[:len(cur)-1]
	}

	return
}

File: cpp/78.cpp
class Solution {
private:
    vector<vector<int>> res;
    vector<int> vec;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(nums, 0);
        return res;
    }
    
    void dfs(vector<int> &nums, int cur){
        if(cur==nums.size()){
            res.push_back(vec);
            return;
        }
        vec.push_back(nums[cur]);
        dfs(nums, cur+1);
        vec.pop_back();
        dfs(nums, cur+1);
    }
};